### PR TITLE
Web form support

### LINKED ISSUE #401: Web forms support
In your talk at the [OKL Berlin](https://www.datenanfragen.de/verein/event/okl-berlin-05-2020/) (at the end around 35min) you have mentioned you would like to support web forms of companies…

Is there an issue tracking this idea yet? Is this idea still on your plan?
If not, then this shall be the issue… :stuck_out_tongue_winking_eye: 

---

So, technically, I think this is very hard, because usually these are cross-origin requests and due to the [same-origin policy](https://developer.mozilla.org/docs/Web/Security/Same-origin_policy) (SOP) these are not allowed. A website would have to [allow CORS requests](https://developer.mozilla.org/docs/Web/HTTP/CORS).

My idea to circumvent this would be a WebExtension (browser add-on), that can offer an API for this project and then submit the requests.

Another problem through maybe that the forms may change fast or even have special CSRF tokens against such things we try to do here… :thinking:

### PR DESCRIPTION
**Pull request by baltpeter**
This will finally fix #401.

The PR isn't quite done yet. TODO:

- [x] In the new generator, the send request modal closes after `finishRequest()` (or, more specifically, after `addRequest()`). I don't know why. This is already not great for emails, but really not good for web forms.
- [x] Update the fallback list of transport mediums.
- [x] Update the company and SVA single pages.
- [x] Ensure the suggest form works with the new field.
- [x] Fix [test error](https://cloud.cypress.io/projects/nug7wk/runs/3933/overview/688d7c59-cb1a-403e-af29-74bc28ebbe7f?reviewViewBy=FAILED).
- [x] Undo the deploy script change.
- [ ] ~~Tests~~

But I'd like to get some feedback on the implementation before I work on those.

### CODE REVIEW

**Comment by baltpeter** [comment]
I have realized that writing tests isn't really possible until we have merged the data PR, so I've deferred that TODO to an issue (#1027).

**Comment by baltpeter** [comment]
I'm looking into the disappearing modal. The problem is that we're fetching the proceedings in `App.tsx`:

https://github.com/datenanfragen/website/blob/32b118a8b175b00aa41c668e8d060df715292385/src/Components/App/App.tsx#L67

When we `finishRequest()` and thus `addRequest()`, a new proceeding is created and the whole app rerenders, which causes the modal to be thrown away.

The amusing part: The only reason we're fetching the proceedings there is for the "don't worry, we still have the old generator" hint. :D

**Comment by baltpeter** [comment]
The problem would disappear if we computed the bool we're interested in directly in the `useProceedingsStore()` call like this:

```ts
const hasUsedOldGenerator = useProceedingsStore(
    (state) =>
        !!Object.values(state.proceedings).find(
            (p) => (getGeneratedMessage(p, 'request')?.date || new Date('9999-12-31')) <= new Date('2022-12-05')
        )
);
```

But @zner0L forbade me from doing that (https://github.com/datenanfragen/website/pull/938#discussion_r992287441). Though, to be fair, he already told me to move this to a store function back then and I just didn't listen. :D And if we do as he said, the problem also disappears…

**Comment by baltpeter** [comment]
This is now ready for merge. It can (and should) be merged before the data changes.

Please rebase merge instead of squashing.

**Comment by baltpeter** [comment]
Schema issue in data: https://github.com/datenanfragen/data/issues/2099

File: src/Components/Generator/WebformModal.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { IntlProvider, Text } from 'preact-i18n';
```
[the remark this answers is not in the document]
Comment (baltpeter) [reply src/Components/Generator/WebformModal.tsx:1]: It's not. With this new addition, we now have three similar but different instances of this code, there's also the original one in `MailtoDropdown`.

I didn't particularly like this either, but I see no other sensible way. The code is very simple, mostly just the markup of which fields to show with which content. But they all need slightly different versions of that. How would you abstract that?

File: src/Components/Generator/RequestForm.tsx [hunk 2]
```diff
@@ -60,14 +61,15 @@ export const RequestForm = (props: RequestFormProps) => {
                         <RequestTypeChooser />
 
                         <TransportMediumChooser
-                            value={transport_medium}
+                            value={transportMedium}
                             onChange={(value) => setTransportMedium(value)}
+                            media={TRANSPORT_MEDIA.filter((m) => m !== 'webform' || current_company?.webform)}
```
[the remark this answers is not in the document]
Comment (baltpeter) [reply src/Components/Generator/RequestForm.tsx:66]: I don't really understand what you're saying. What allowlist?

File: src/i18n/de.json [hunk 3]
```diff
@@ -380,6 +382,10 @@
         "mailto-dropdown-explanation": "Wie möchtest Du die E-Mail verschicken?",
         "mailto-dropdown-warning": "*Dadurch werden die Daten Deiner Anfrage an den jeweiligen Anbieter weitergegeben.",
         "copymanually-explanation": "Hier kannst Du den Betreff, Empfänger und Text Deiner generierten E-Mail von Hand kopieren.",
+        "webform-modal-explanation": "Hier kannst Du den Betreff und Text Deiner generierten Anfrage kopieren, um sie in das Web-Formular einzufügen.",
```
[the remark this answers is not in the document]
Comment (baltpeter) [reply src/i18n/de.json:385]: But this translation is only for the advanced generator. It's possible that the user chose to send the request via web form, even though the default was email.

File: src/Components/Generator/TransportMediumChooser.tsx [hunk 4]
```diff
@@ -17,14 +18,14 @@ export function TransportMediumChooser(props: TransportMediumChooserProps) {
             <fieldset className="request-transport-medium-chooser label-only-fieldset">
                 <legend>{props.label ? props.label : <Text id="request-transport-medium" />}</legend>
                 <div className="radio-group">
-                    {TRANSPORT_MEDIA.map((transport_medium) => (
+                    {(props.media || TRANSPORT_MEDIA.filter((m) => m !== 'webform')).map((transportMedium) => (
```
[the remark this answers is not in the document]
Comment (baltpeter) [reply src/Components/Generator/TransportMediumChooser.tsx:21]: #1023

**Review by zner0L — CHANGES_REQUESTED** [review]
Ok, I managed to get a test environment going that was at least good enough for testing the UX and all in all I really like it. Still, I have a few notes on it:

- [ ] I think it would be really useful to copy the message text to the clipboard if someone clicks the link to the form, so that you don't have to switch tabs that often. The 1Password Extension does this as well (at least on iPad OS) and I find it really convenient.
- [ ] This is not blocking but maybe we sould mark webform companies in the search as we do with mail/fax only companies as well?
- [ ] In the new generator flow, there definitely needs to be more expanation as to why there is a link to a webform suddenly. I don't think people necessarily expect to go to another website, when they click that button.

Other than these, I am very happy with how this turned out.

**Comment by baltpeter** [comment]
> I think it would be really useful to copy the message text to the clipboard if someone clicks the link to the form, so that you don't have to switch tabs that often. The 1Password Extension does this as well (at least on iPad OS) and I find it really convenient.

@mal-tee also suggested that (discussion is unfortunately in the Matrix chat, not here). I agree that such a feature would be useful, but I really don't want to just overwrite the user's clipboard. I have opened #1025 to track the discussion.

**Comment by baltpeter** [comment]
> This is not blocking but maybe we sould mark webform companies in the search as we do with mail/fax only companies as well?

I've opened #1032 to track that. I have no strong opinion on that.

**Comment by baltpeter** [comment]
> In the new generator flow, there definitely needs to be more expanation as to why there is a link to a webform suddenly. I don't think people necessarily expect to go to another website, when they click that button.

Not sure whether that is a problem specifically for web forms. Maybe I'm wrong but I feel like the explanation for emails isn't exactly better (and neither opening a tab to a web form nor mailto links are exactly common UX paradigms that users expect).

But if you have a suggestion for an explanation, do tell.

File: src/store/proceedings.ts [hunk 5]
```diff
@@ -158,12 +161,19 @@ const proceedingsStore = persist<ProceedingsState>(
                             window.ON_PROCEEDING_STATUS_CHANGE?.(state.proceedings[ref], oldStatus);
                         }
                     }
+
+                    state._hasUsedOldGenerator = !!Object.values(state.proceedings).find(
```
[the remark this answers is not in the document]
Comment (baltpeter) [reply src/store/proceedings.ts:165]: Sure. :D

**Review by zner0L — APPROVED** [review]
Thanks for the changes! I move my last request to an issue as well, so this is good to go.

**Pull request merged**